File: MIDTERM-LAB-3/MP15/MP15_ExportFirst50.java
// ============================================================
// MP15 - Export First 50 Rows to CSV
// Programmer: OBNIAL, MARIUS RYZER C.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program reads the CSV dataset, takes the
//              first 50 data rows (plus the header), and saves
//              them to a new CSV file called "output_mp15.csv".
// ============================================================

import java.io.*;        // For reading and writing files
import java.util.*;      // For ArrayList

public class MP15_ExportFirst50 {

    public static void main(String[] args) {

        // Scanner is used to read keyboard input
        Scanner scanner = new Scanner(System.in);

        // Step 1: Ask the user where the CSV file is located
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        // Step 2: Ask the user where to save the output file
        System.out.print("Enter output file name (e.g., output_mp15.csv): ");
        String outputPath = scanner.nextLine().trim();

        // If no output name was given, use a default name
        if (outputPath.isEmpty()) {
            outputPath = "output_mp15.csv";
        }

        // 'headerLine' stores the header row to include at the top of the output
        String headerLine = null;

        // 'exportedRows' stores the data rows we will write to the output file
        ArrayList<String> exportedRows = new ArrayList<>();

        // 'dataRowCount' counts how many actual data rows we have collected
        int dataRowCount = 0;

        // 'headerFound' tracks whether we've found the real column header
        boolean headerFound = false;

        // Maximum number of rows we want to export
        final int MAX_ROWS = 50;

        // --- Step 3: Read the original CSV file ---
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line; // 'line' holds the current line being read

            while ((line = br.readLine()) != null) {

                // Skip completely blank lines
                if (line.trim().isEmpty()) continue;

                // Split the line into column values
                String[] columns = line.split(",", -1);

                // --- Find the header row ---
                // Look for the line that starts with "Candidate"
                if (!headerFound && columns[0].trim().equalsIgnoreCase("Candidate")) {
                    headerFound = true;   // Mark header as found
                    headerLine = line;    // Save the header line
                    continue;             // Don't add it to data yet
                }

                // --- Collect data rows (up to MAX_ROWS = 50) ---
                if (headerFound && dataRowCount < MAX_ROWS) {
                    exportedRows.add(line); // Add this row to our export list
                    dataRowCount++;         // Increase the data row counter

                    // Once we have 50 rows, stop reading
                    if (dataRowCount >= MAX_ROWS) break;
                }
            }

        } catch (FileNotFoundException e) {
            System.out.println("ERROR: Input file not found. Check the file path.");
            scanner.close();
            return;
        } catch (IOException e) {
            System.out.println("ERROR: Could not read the file. " + e.getMessage());
            scanner.close();
            return;
        }

        // --- Step 4: Write the output CSV file ---
        // PrintWriter creates/overwrites a file and lets us write text to it
        try (PrintWriter pw = new PrintWriter(new FileWriter(outputPath))) {

            // Write the header row first (if we found one)
            if (headerLine != null) {
                pw.println(headerLine);
            }

            // Write each of the collected data rows
            for (String row : exportedRows) {
                pw.println(row);
            }

            // If PrintWriter had any errors while writing:
            if (pw.checkError()) {
                System.out.println("ERROR: Something went wrong while writing the file.");
                scanner.close();
                return;
            }

        } catch (IOException e) {
            System.out.println("ERROR: Could not write output file. " + e.getMessage());
            scanner.close();
            return;
        }

        // --- Step 5: Display confirmation ---
        System.out.println("\n============================================================");
        System.out.println("          MP15 - EXPORT FIRST 50 ROWS TO CSV");
        System.out.println("============================================================");
        System.out.println("  Input File  : " + filePath);
        System.out.println("  Output File : " + outputPath);
        System.out.println("  Rows Exported (excluding header) : " + dataRowCount);
        System.out.println("------------------------------------------------------------");

        // Preview the first 5 exported rows as a sample
        System.out.println("  Preview (first 5 exported rows):");
        for (int i = 0; i < Math.min(5, exportedRows.size()); i++) {
            System.out.println("    " + exportedRows.get(i));
        }

        System.out.println("------------------------------------------------------------");
        System.out.println("  Export complete! File saved to: " + outputPath);
        System.out.println("============================================================");

        scanner.close(); // Always close scanner at the end
    }
}

File: MIDTERM-LAB-3/MP15/MP15_ExportFirst50.js
// ============================================================
// MP15 - Export First 50 Rows to CSV
// Programmer: OBNIAL, MARIUS RYZER C.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program reads the CSV dataset, takes the
//              first 50 data rows plus the header, and writes
//              them to a new CSV output file.
// ============================================================

// 'fs' = Node.js built-in module for file reading and writing
const fs = require("fs");

// 'readline' = Node.js module for reading keyboard input (terminal)
const readline = require("readline");

// Set up the terminal input/output interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Step 1: Ask the user for the input CSV file path
rl.question("Enter the CSV dataset file path: ", function (filePath) {

  // Step 2: Ask where to save the output file
  rl.question(
    "Enter output file name (e.g., output_mp15.csv): ",
    function (outputPath) {

      outputPath = outputPath.trim();

      // If no output name given, use a default
      if (outputPath === "") {
        outputPath = "output_mp15.csv";
      }

      // Call the main export function
      exportFirst50(filePath, outputPath);

      rl.close(); // Done with input, close the interface
    }
  );
});

// ============================================================
// FUNCTION: exportFirst50
// PURPOSE : Reads the input CSV, extracts the header + first
//           50 data rows, writes them to a new CSV file,
//           and shows a preview and summary.
// PARAMETERS:
//   filePath   - path to the input CSV file
//   outputPath - path/name for the output CSV file
// ============================================================
function exportFirst50(filePath, outputPath) {

  // --- Step 1: Check if the input file exists ---
  if (!fs.existsSync(filePath)) {
    console.log("ERROR: File not found. Please check the file path.");
    return;
  }

  // --- Step 2: Read the entire file content as text ---
  const fileContent = fs.readFileSync(filePath, "utf8");

  // --- Step 3: Split the text into individual lines ---
  const lines = fileContent.split("\n");

  // --- Variables ---
  let headerLine = null;       // Stores the header row text
  let exportedRows = [];       // Stores the 50 data rows to export
  let headerFound = false;     // Did we find the header row yet?
  const MAX_ROWS = 50;         // We want exactly 50 data rows

  // --- Step 4: Loop through each line ---
  for (let i = 0; i < lines.length; i++) {

    let line = lines[i].trim(); // Current line without leading/trailing spaces

    // Skip blank lines
    if (line === "") continue;

    // Parse the line into column values
    let columns = parseCSVLine(line);

    // --- Find the header row ---
    // It's the row where the first column says "Candidate"
    if (!headerFound && columns[0].trim().toLowerCase() === "candidate") {
      headerFound = true; // Header located!
      headerLine = line;  // Save it for the output file
      continue;           // Don't add to data rows
    }

    // --- Collect data rows after the header (up to 50) ---
    if (headerFound && exportedRows.length < MAX_ROWS) {
      exportedRows.push(line); // Add this row to our export list

      // Once we hit 50 rows, stop collecting
      if (exportedRows.length >= MAX_ROWS) break;
    }
  }

  // --- Step 5: Build the output content ---
  // We join header + data rows into one big string, each on its own line
  let outputLines = [];

  if (headerLine !== null) {
    outputLines.push(headerLine); // First line = the header
  }

  // Add all the collected data rows
  exportedRows.forEach(function (row) {
    outputLines.push(row);
  });

  // Join all lines with newline characters to make the final file content
  let outputContent = outputLines.join("\n");

  // --- Step 6: Write the output to a new file ---
  // 'fs.writeFileSync' creates/overwrites a file with the given content
  // 'utf8' means write it as normal readable text
  try {
    fs.writeFileSync(outputPath, outputContent, "utf8");
  } catch (err) {
    console.log("ERROR: Could not write the output file. " + err.message);
    return;
  }

  // --- Step 7: Display results and preview ---
  console.log("\n============================================================");
  console.log("          MP15 - EXPORT FIRST 50 ROWS TO CSV");
  console.log("============================================================");
  console.log("  Input File  : " + filePath);
  console.log("  Output File : " + outputPath);
  console.log("  Rows Exported (excluding header) : " + exportedRows.length);
  console.log("------------------------------------------------------------");

  // Show a preview of the first 5 exported rows
  console.log("  Preview (first 5 exported rows):");
  let previewCount = Math.min(5, exportedRows.length); // Show at most 5
  for (let i = 0; i < previewCount; i++) {
    console.log("    " + exportedRows[i]);
  }

  console.log("------------------------------------------------------------");
  console.log("  Export complete! File saved to: " + outputPath);
  console.log("============================================================");
}

// ============================================================
// FUNCTION: parseCSVLine
// PURPOSE : Splits a single CSV line into an array of values,
//           correctly handling quoted fields with commas inside.
// EXAMPLE : "Obnial, Marius",Student,,Python
//           → ["Obnial, Marius", "Student", "", "Python"]
// PARAMETER: line - one line of text from the CSV
// RETURNS: array of string values
// ============================================================
function parseCSVLine(line) {
  let result = [];      // Array to store parsed values
  let current = "";     // Current field being built
  let inQuotes = false; // Flag: are we inside a quoted field?

  for (let i = 0; i < line.length; i++) {
    let char = line[i];

    if (char === '"') {
      inQuotes = !inQuotes; // Toggle in/out of quoted section
    } else if (char === "," && !inQuotes) {
      result.push(current); // End of a field — save it
      current = "";          // Reset for next field
    } else {
      current += char; // Add character to current field
    }
  }

  result.push(current); // Push the last field (no trailing comma)
  return result;
}